packages/components/src/text/Text.styles.js

```diff
@@ -0,0 +1,14 @@
+/**
+ * External dependencies
+ */
+import styled from '@emotion/styled';
+
+/**
+ * Internal dependencies
+ */
+import { text } from './mixins';
+
+export const Text = styled.p(
```

Comment:
@jameslnewell re: the base `Text` component.

Would love your thoughts on this. In my recent `Card` component, I wrapped the styled components with another component to provide traditional CSS classNames:

https://github.com/WordPress/gutenberg/blob/master/packages/components/src/card/body.js#L22

The idea for that was to enable folks to adjust styles (if they wanted to) with regular CSS. I also found that having these regular classes made it easier for debug (via browser inspector tools). Lastly, I found that it made it more accessible for folks unfamiliar with React/CSS-in-JS to navigate to debug/contribute.

Reply:
Any thoughts about how to make this `Text` component more flexible to support other HTML tags like `h1`...`h6`, `span`, `strong`, `em`, etc.?

Reply:
That should be possible with Emotion's `as` prop 😊 
https://emotion.sh/docs/styled#as-prop

```
<Text as='h2'>Hi!</Text>
```

Reply:
Sweet:

```js
import styled from '@emotion/styled'

const Button = styled.button`
  color: hotpink;
`

render(
  <Button
    as="a"
    href="https://github.com/emotion-js/emotion"
  >
    Emotion on GitHub
  </Button>
)
```

I like that example, it's explicit. We do some magic in Gutenberg when you use `Button` and pass `href` as prop to achieve the same goal.

Reply:
> Would love your thoughts on this. In my recent Card component, I wrapped the styled components with another component to provide traditional CSS classNames...

I'd prefer to not encourage consuming developers to rely on markup/classnames as that exposes the internals of a component which should be free to change. I'd prefer to encourage extensibility through composibility (i.e. mixins or smaller more focussed components). With `styled` there's nothing stopping the consumer from applying their own styles.

e.g.

Adjusting styles with traditional methods:
```jsx
<Text className="is-underlined"/>
```

Adjusting styles with CSS-in-JS:
```jsx
const TextWithUnderline = styled(Text)`
  text-decoration: underline;
`;

<TextWithUnderline/>
```

Reply:
@gziolo Yep `as` is the goto for using a more semantic element. 

However, I think for `Button`, the current implementation is preferred over the use of `as` because when a `Button` component is rendered as an anchor the component also needs to apply a bunch of other attributes and behaviour e.g. `role="button"` and handling `space`

Reply:
One important aspect here is that our consuming developpers are not exclusively using React. In WordPress we have style plugins, we have plugins relying on DOM. Granted, these are not great ways to extend the UI but I'm not certain we can make the decision today to remove classNames.

Reply:
@youknowriad I thought the goal of `@wordpress/components` was to facilitate the app-like UI/UX for editing blocks, not the actual markup rendered by the block on the "frontend" (therefore classnames aren't particularly important)?

Or are you suggesting style plugins should be able to style the editor interface as well as the "frontend" experience?

Reply:
Yes, I'm not really talking about frontend here. You can expect plugin developpers to try to override styles for the editor UI... Not that it's the best way to do it but something they'll 100% try to do.

Reply:
Is overriding editor styles something we're wanting to enable and encourage developers to do? 

If that's the case we'll always require classnames, but for consumers of the components (e.g. block developers), I expect the reliance on markup structure and classnames will dramatically decrease as the components become smaller and more composable and have built-in-support for theming.

Reply:
> Is overriding editor styles something we're wanting to enable and encourage developers to do?

That's a fair question. Personally, I think the answer is yes, but ideally, it should be done in a way that does not involve CSS overrides (that feel hacky). The Theming solution from Emotion would be an example of a good way to do this (as you pointed out @jameslnewell )

But... to @youknowriad 's concerns... that workflow would only be accessible to the React developers. And since our current CSS-in-JS (Emotion) setup is internal and subject to change, it's a little bit risky.

Pure CSS-in-JS (although ideal) may be too early at this point. Having classNames available (like the other components) enables compatibility with the workflow of today while setting us up for the styling solution we want for the future.

Reply:
There is no existing `Text` component today so not adding classes on this component won't affect the existing workflow for a style plugin today.

The consuming developer is free to apply their own classname or override the styles of this component e.g.
```jsx
<Text variant="..." className="rainbow-text"/>
// OR
const RainbowText = styled(Text)`
  ...
`;
// OR
const RainbowText = styled.p`
  ${ text( { variant: 'text.large' } ) }
`;
```

There's no reason a style plugin can't continue using the existing classnames that applied to components today e.g. https://github.com/WordPress/gutenberg/pull/18652/files#r348899237